webob/response.py

```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
         # Response with empty body, such as Response(status='204 No Content')
         # without the default content_type being set
 
-        if (
-            self.content_type is None and
-            headerlist is None and
-            _code_has_body(self.status_code)
-        ):
-            self.content_type = content_type
+        encoding = None
+        code_has_body = (
+            self._status[0] != '1' and
+            self._status[:3] not in ('204', '205', '304')
+        )
+        if headerlist is None and code_has_body:
+            content_type = content_type or self.default_content_type
 
-        # Set up the charset
-        #
-        # In contrast with the above, if a charset is not set but there is a
-        # content_type we will set the default charset if the content_type
-        # allows for a charset.
+            # Check if content_type is set because default_content_type could
+            # be None, in which case there is no content_type
+            if content_type:
+
+                # Set up the charset, if the content_type doesn't already have one
 
-        if self.content_type:
-            if not self.charset and charset is not _marker:
-                self.charset = charset
-            elif not self.charset and self.default_charset:
-                if _content_type_has_charset(self.content_type):
-                    self.charset = self.default_charset
+                has_charset = 'charset=' in content_type
+
+                if not has_charset and charset is not _marker and charset is not None:
+                    encoding = charset
+                    content_type += '; charset=' + charset
+                elif not has_charset and charset is _marker and self.default_charset:
+                    # Optimize for the default_content_type case, avoid a function call
+                    if (
+                        content_type is self.default_content_type or
+                        _content_type_has_charset(content_type)
+                    ):
+                        encoding = self.default_charset
+                        content_type += '; charset=' + self.default_charset
+
+            self._headerlist.append(('Content-Type', content_type))
 
         # Set up conditional response
         if conditional_response is None:
             self.conditional_response = self.default_conditional_response
         else:
             self.conditional_response = bool(conditional_response)
 
-        # Set up app_iter
-        if app_iter is None:
+        # Set up app_iter if the HTTP Status code has a body
+        if app_iter is None and code_has_body:
             if isinstance(body, text_type):
-                encoding = self.charset
+                # Fall back to trying self.charset if encoding is not set. In
+                # most cases encoding will be set to the default value.
+                encoding = encoding or self.charset
```

Comment:
`body` has accepted only `bytes` (or it's equivalent `str` in Py2 for a VERY long time): https://github.com/Pylons/webob/blob/0.8/webob/__init__.py#L1445

It would be an API asymmetry to change that, because `r.body` will always return bytes, which may be unexpected. I feed it text, I get back bytes from the same property?

That's what `.text` is for.

The fact that `Response(body=u'foo')` works is that in 0.9 `__init__` was updated to call `unicode_body`: https://github.com/Pylons/webob/blob/0.9/webob/__init__.py#L1364 when the "body" provided was unicode. I assume it was a convenience thing, because `.body` was not updated to match. `.unicode_body` sometime in its history got renamed to `.text`, and instead of calling `self.unicode_body` vs `self.body` the features from it were inlined into `__init__` probably for speed reasons, which brings us back to where we are today.

Unfortunately the name "body" is being re-used for something that accepts both `bytes`/`str` and `unicode`.

----

That being said, falling back to `default_body_encoding` seems like it could be a sane thing to do, and matches what was done in the past.

Side note: Ian at one point did only set the default `charset` on text types:

https://github.com/Pylons/webob/blob/0.9/webob/__init__.py#L1357

Somewhere between that version and 1.6.x it was lost, and thus if you didn't explicitly provide a `charset` providing `u'foo'` as a `body` parameter would have failed due to not having a valid charset.